### PR TITLE
Multiple supporting file upload

### LINKED ISSUE #1158: Uploaded file name not appearing on navigating forwards and back
**Describe the bug**
The name of the uploaded file is not appearing in the dropzone molecule if you navigate forwards and then backwards in the wizard.

**Steps To Reproduce**
1. Navigate to Staging
2. On the files step, upload a manuscript
3. Observe the filename is shown along with success message
4. Click 'Next'
5. Click 'Back'
6. Filename is now longer shown

**Expected behavior**
Filename should be retained after a manuscript has been successfully uploaded.

### PR DESCRIPTION
**Pull request by Cooryd**
#### Background

Supporting file upload is an MVP requirement for the first release of xpub. The user should be able to upload up to 10 supporting documents after they have uploaded their manuscript file.

What does this PR do?

This branch implements a basic multi file upload component for uploading supporting files as well as making the required edits to the manuscript upload/ delete mutations to prevent these parts breaking.

Worked in the inclusion of file name on manuscript upload in the PR when returning to the upload page.

Also includes `babel-pollyfil` in styleguide compilation.

closes #1158

#### How has this been tested?
Basic unit test which needs extending. Testing will be addressed in #1163

#### UI Changes

- [ ] Has been reviewed against Designs
This is a minimum working component so will need to be re-looked at in future
- [x] Necessary updates to styleguide

### CODE REVIEW

**Comment by hdrury1** [comment]
FYI I've been unable to take a look at this in the review environment as I get the following error when attempting to upload a manuscript.

`Unhandled promise rejection Error: "GraphQL error: Access Denied"
	t http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:28:204114
	next http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:98:84836
	y http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:106822
	b http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:107328
	value http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:107881
	y http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:106822
	b http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:107328
	value http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:90:107881
	s http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:98:149339
app.fba6af35da1a6ae1adaf.js:66:257602
I/</t<
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:66:257602
e.exports
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:47:33138
I/<
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:66:257479
e.exports
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:47:30939
<anonymous>
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:28:169491
y
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:28:169349
b
http://35.172.178.78:32075/assets/app.fba6af35da1a6ae1adaf.js:28:169369
`

**Comment by Cooryd** [comment]
@hdrury1 The styleguide should now be working, the component is called `SupportingUpload`. I've got it randomly erroring for the purpose of demonstrating all states.

**Comment by hdrury1** [comment]
Managed to run this locally and it's looking great. One key piece of functionality missing that I really do think we'd need is the ability to clear all uploaded files. This wouldn't need to be on an individual file basis- it would be enough for a user to be able to remove all files at once and start again.

**Comment by diversemix** [comment]
@Cooryd did you push the update to do the delete all? If so @hdrury1 this is ready to test!

**Comment by Cooryd** [comment]
I had fixed the bug I was working on then completely forgot about the delete all feature. Shouldn’t be a big piece of work and should be done early tomorrow.

**Comment by diversemix** [comment]
Ah no worries. You did some great work today with Semra... I misunderstood!  :+1:

**Comment by diversemix** [comment]
@hdrury1 What's the status on this now?

**Comment by Cooryd** [comment]
@hdrury1  - This should now be ready to look over with the delete option (I would say we need to add a  popconfirm to this like with the manuscript delete option)

**Comment by hdrury1** [comment]
Following discussions with Chris:

1) Validation text should only appear once 10 files have been uploaded
2) In the event the application has to cap files (e.g. user uploads 6 files, followed by 7 and the system only saves the first 4) the validation text should alert the user that the last 3 files are over the 10 file limit
3) Add a header (Supporting Files) above the table of files once at least one supporting file has been uploaded

**Comment by diversemix** [comment]
There are two technical issues:
- The client queue for files means that we should prevent navigating away - disable back and next buttons.
- The auto save feature means any changes to the cover letter (while an upload is in progress) do not get saved ... as the server keeps the original Manuscript object in memory so overwrites on save. .... Talked about a improvement in the ORM to throw an exception on `save()` if saving an old version of the object. Also this should be fixed by refreshing the in memory object before saving.

**Comment by Cooryd** [comment]
Need to look at mobile controls (currently hidden because of problems with getting the layout right)

File: app/components/pages/SubmissionWizard/steps/Files/ManuscriptUpload.js [hunk 1]
```diff
@@ -213,7 +213,7 @@ class ManuscriptUpload extends React.Component {
               conversion={conversion}
               dropzoneOpen={() => dropzoneRef.open()}
               errorMessage={this.state.errorMessage}
-              fileName={this.fileName}
+              fileName={this.fileName || fileName}
```
[the remark this answers is not in the document]
Comment (Cooryd) [reply app/components/pages/SubmissionWizard/steps/Files/ManuscriptUpload.js]: If the component has an internal fileName set then this displays over a prop value (which would be the case on a new upload)

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    const successfullyUploadedFiles = this.state.files.filter(
+      file => !file.error,
+    )
+    return (
+      <React.Fragment>
+        <StyledDropzone
+          maxSize={config.fileUpload.maxSizeMB * 1e6}
+          onDrop={droppedFiles => {
+            let files = droppedFiles
+            if (files.length > 10 - successfullyUploadedFiles.length) {
```
Comment (diversemix) [inline app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: Magic number here ... can we put in props say via config?

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  render() {
+    let dropzoneRef
+    const { hasManuscript, removeFiles } = this.props
+    const successfullyUploadedFiles = this.state.files.filter(
+      file => !file.error,
+    )
+    return (
+      <React.Fragment>
+        <StyledDropzone
+          maxSize={config.fileUpload.maxSizeMB * 1e6}
```
Comment (diversemix) [inline app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: Where do we use `maxSize` ?

File: app/components/pages/SubmissionWizard/steps/Files/index.js [hunk 4]
```diff
@@ -95,12 +123,22 @@ const FilesPageContainer = ({
   uploadData = {},
   uploadLoading,
 }) => (
-  <Mutation mutation={UPLOAD_MUTATION}>
+  <Mutation mutation={UPLOAD_MANUSCRIPT_MUTATION}>
     {(uploadFile, { loading, error: uploadError }) => {
       const fieldName = 'files'
```
Comment (diversemix) [inline app/components/pages/SubmissionWizard/steps/Files/index.js]: Not sure if `fieldName` here is very descriptive ... I'd use `const submissionFiles = values['files]`

File: app/components/pages/SubmissionWizard/steps/Files/ManuscriptUpload.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cooryd) [reply app/components/pages/SubmissionWizard/steps/Files/ManuscriptUpload.js]: I've gone with 'droppedFileName', seems descriptive enough of what its storing

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    100% {
+      transform: rotate(360deg);
+    }
+  }
+`
+
+function* fileUploadGenerator(files, uploadFile) {
+  for (let fileIndex = 0; fileIndex < files.length; ) {
+    yield {
```
Comment (damnhipster) [inline app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: 👍

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+          .then(data => {
+            this.updateFileState(result.value.fileId, {
+              success: true,
+              loading: false,
+            })
+          })
+          .catch(() => {
+            this.updateFileState(result.value.fileId, {
+              error: true,
+              loading: false,
```
Comment (damnhipster) [inline app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: What does this error state look like?

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    loop(iterator.next())
+  }
+
+  updateFileState = (fileId, newState) => {
+    const newFilesState = [...this.state.files]
+    const fileIndex = newFilesState.findIndex(file => file.id === fileId)
+    if (fileIndex > -1) {
+      newFilesState[fileIndex] = { ...newFilesState[fileIndex], ...newState }
+      this.setState({ files: newFilesState })
```
Comment (damnhipster) [inline app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: 👍

File: app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js (hunk 6 above)
Comment (Cooryd) [reply app/components/pages/SubmissionWizard/steps/Files/SupportingUpload.js]: red border and the same error icon as the manuscript upload failing

File: server/xpub-model/entities/file/index.js [hunk 8]
```diff
@@ -51,15 +51,23 @@ class File extends BaseModel {
   }
 
   static async find(id) {
-    const [file] = await this.query().where({
-      manuscript_id: id,
+    const file = await this.query().where({
+      id,
```
Comment (damnhipster) [inline server/xpub-model/entities/file/index.js]: I think `pubsweet-model` has a `find` method on the base class which has this same behaviour, so we may not need to override it anymore.

File: server/xpub-server/entities/manuscript/resolvers/removeSupportingFiles.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+const { Manuscript, File, User } = require('@elifesciences/xpub-model')
+const logger = require('@pubsweet/logger')
+
+async function removeSupportingFiles(_, { id }, { user }) {
+  const userUuid = await User.getUuidForProfile(user)
+  const manuscript = await Manuscript.find(id, userUuid)
+
+  manuscript.files.forEach((file, index) => {
+    if (file.type === 'SUPPORTING_FILE') {
+      manuscript.files.splice(index, 1)
```
Comment (damnhipster) [inline server/xpub-server/entities/manuscript/resolvers/removeSupportingFiles.js]: `splice` may cause an issue with the iterator.

```
> const arr  = [1,2,3,4,5,6]
> arr.forEach((el, index) => {
... arr.splice(index,1)
... console.log(arr)
... })
[ 2, 3, 4, 5, 6 ]
[ 2, 4, 5, 6 ]
[ 2, 4, 6 ]
>
```

File: webpack/webpack.development.config.js [hunk 10]
```diff
@@ -15,6 +15,7 @@ module.exports = [
       app: [
         'react-hot-loader/patch',
         'webpack-hot-middleware/client',
+        'babel-polyfill',
```
[the remark this answers is not in the document]
Comment (Cooryd) [reply webpack/webpack.development.config.js]: generators require a pollyfil to work. Was already included in `production` but missing in `development` and `styleguide`

**Comment by damnhipster** [comment]
Will tests the changes in the morning.

**Comment by giorgiosironi** [comment]
After:
- [Hem's IAM user being added to the K8S cluster whitelist](https://github.com/elifesciences/terraform-eks/commit/13817a1a828cad26fe23ff1aee641dfddc6a2ada)
- the [`Xpub` IAM group](https://console.aws.amazon.com/iam/home?region=us-east-1#/groups/Xpub) now having the [same policy as Jenkins](https://console.aws.amazon.com/iam/home?region=us-east-1#/policies/arn:aws:iam::512686554592:policy/AmazonKMSXpubMinimalUserAccess$jsonEditor)

you should now be able to setup `kubectl` and use it to access any pod.

**Comment by damnhipster** [comment]
> After:
> 
> * [Hem's IAM user being added to the K8S cluster whitelist](https://github.com/elifesciences/terraform-eks/commit/13817a1a828cad26fe23ff1aee641dfddc6a2ada)
> * the [`Xpub` IAM group](https://console.aws.amazon.com/iam/home?region=us-east-1#/groups/Xpub) now having the [same policy as Jenkins](https://console.aws.amazon.com/iam/home?region=us-east-1#/policies/arn:aws:iam::512686554592:policy/AmazonKMSXpubMinimalUserAccess$jsonEditor)
> 
> you should now be able to setup `kubectl` and use it to access any pod.

Yes, I've managed to access the cluster.

**Pull request merged**